short_alloc: if allocation still fits exactly into arena, it should be used

ajneu opened the pull request:
Thanks for `short_alloc`! (I've been experimenting with using it with containers without using the freestore, which is nicely efficient (and a requirement in some safety-critical software, or so I'm told)).

I noticed this:
Overlarge alignments can cause the allocation to still fits exactly,
yet cause the arena's `ptr_` to jump out of the arena.
But the allocation is then still valid, even though `ptr_` ends up outside
the arena, since the new position of `ptr_` is only for the _next_ allocation!

Here's a code example. 
It shows non-zero memory, when using the existing version, 
but zero-memory-usage (as desired) when using the code from the pull-request.

``` cpp
#include "short_alloc.h"
#include <iostream>
#include <new>
#include <vector>

// Replace new and delete just for the purpose of demonstrating that
//  they are not called.

std::size_t memory = 0;
std::size_t alloc = 0;

void* operator new(std::size_t s) //throw(std::bad_alloc)
{
  memory += s;
  ++alloc;
  return std::malloc(s);
}

void  operator delete(void* p) noexcept//throw()
{
  --alloc;
  std::free(p);
}

void operator delete(void* ptr, std::size_t) noexcept //throw()
{
    ::operator delete(ptr);
}

void memuse()
{
    std::cout << "memory = " << memory << '\n';
    std::cout << "alloc = " << alloc << '\n';
}

template <class T, std::size_t BufSize = 3*sizeof(T), std::size_t Alig = alignof(T)>
using SmallVector = std::vector<T, short_alloc<T, BufSize, Alig>>;

template <class T, std::size_t BufSize = 3*sizeof(T), std::size_t Alig = 2*alignof(T)> // increase alignment requirement
using SmallVectorXXX = std::vector<T, short_alloc<T, BufSize, Alig>>;

int main()
{
    {
        SmallVector<int>::allocator_type::arena_type a;
        SmallVector<int> v{{1, 2, 3}, a}; // exactly one allocation is made, for exactly 3 ints == 12 bytes
    }
    memuse();

    {
        SmallVectorXXX<int>::allocator_type::arena_type a;
        SmallVectorXXX<int> v{{1, 2, 3}, a}; // exactly one allocation is made, for exactly 3 ints == 12 bytes
        // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        // we jump arena's ptr_ forward a bit more (out of the arena) because of the large alignment
    }
    memuse();
}
```

ajneu committed:
memuse: if allocation still fits exactly into arena, it should be used

Overlarge alignments can cause the allocation to still fits exactly,
yet cause the arena's ptr_ to jump out of the arena.
But the allocation is then still valid, even though ptr_ ends up outside
the arena, since the new position of ptr_ is only for the *next* allocation!

(Also: warnings removed, when compiling with -Wall -Wextra -Werror )

HowardHinnant commented:
Thanks!  I agree you have uncovered a bug.

Currently I'm undecided if the best way to fix this bug is to allow the last allocation to be less than `alignment` (as you propose), or to `static_assert` that `size` is a multiple of `alignment`.

HowardHinnant commented:
I'm leaning towards the `static_assert`.  Detecting and dealing with a last allocation of size smaller than alignment is going to cost cycles, and this thing is all about boosting performance down to a few cycles.

ajneu committed:
fix deallocate, for cases where operator new allocated a p that is less than buf_

ajneu commented:
> What if `operator new` allocated a pointer p that compares less than `buf_ - n`?

Ah yes, right! I missed that case:
`deallocate`, would then fail to delete, if `p < buf_`.

There's a fix for that, in the next commit (see just before this comment).

ajneu commented:
By the way, I've just noticed: while the `arena` nicely enforces alignment; the fallback (once the `arena` is full), which calls `new`, does not enforce any alignment.

I wonder if there is a different way of using `new` that allows alignment requirements?
- Looking here http://stackoverflow.com/a/16510895/5106243 we see `new` (at least in its normal usage) only guarantees a constant maximum alignment (`alignof(std::max_align_t)`) for every call to `new`.

Hence I suggest adding something like

``` cpp
static_assert(alignment <= alignof(std::max_align_t), "you've chosen an alignment that is larger than alignof(std::max_align_t), and cannot be guaranteed by normal operator new");
```

**right before** the call to `new`.

.
.
.
.
.
.
Of course getting `new` to do user-specified alignments would be better, but the relevant function is

``` cpp
// template...
char* arena<N, alignment>::allocate(std::size_t n);
```

where `n` is not `constexpr`, so I don't think we can use `std::aligned_storage` or special structs. hmmm

ajneu committed:
static_assert to warn the user that operator new cannot do larger alignments
than alignof(std::max_align_t)

ajneu commented:
> Currently I'm undecided if the best way to fix this bug is to allow the last allocation to be less than `alignment` (as you propose), or to `static_assert` that `size` is a multiple of `alignment`.

Well... here's an example where that `static_assert` might not do what one wants:
https://github.com/ajneu/short_alloc_experiments

Note that the code is "perfectly legal" (I hope), but I've commented the `static_assert` in [this line](https://github.com/ajneu/short_alloc_experiments/blob/c3c543e68733077e46619d0bc3122153a1b17ae8/main.cpp#L50), so that it can compile.

HowardHinnant commented:
:+1:  on static asserting that `alignment <= alignof(std::max_align_t)`!
I'll be digesting the rest more slowly...
Thanks!

HowardHinnant commented:
In your example, `sizeof(vec) == 48`  whether your `NUM_EL == 2` or `NUM_EL == 8`.  In other words, you've allocated space up to the next multiple of your alignment.  It seems a shame to then ignore that allocation.

ajneu commented:
> In your example, sizeof(vec) == 48 whether your NUM_EL == 2 or NUM_EL == 8.

Yes. But the real question is: Why is this so?

> In other words, you've allocated space up to the next multiple of your alignment. 

Nope, not me. **Your** short_allocator did this, because that's what the C/C++ code does with arrays that are aligned (overaligned) within structs.
I just encountered it a few days ago (on an unrelated project) and was **quite surprised myself**!!!

Example: What do you think, will be the output of this code?

``` cpp
#include <iostream>

struct AAA {
   alignas(8) char arr[1];
};

int main(void) {
    std::cout << sizeof(AAA) << std::endl; // surprise?
}
```

Maby 1, right?
Well, that was my surprise. It's not!
I don't know where in the standard they mandate the behaviour or why the compiler does it like that.
But what we have here, are 7 wasted bytes!!!

In the above example, we'd think the important thing is to properly place `AAA` on to a 8byte-boundary (64 bit boundary). It does, but it also has this side-effect, of wasting the ensuing 7 bytes.

> It seems a shame to then ignore that allocation.

I agree. But I'm the wrong person to fix it. ;)
I wonder why compilers do that, and where in the standard that wasteful behaviour is actually specified.

.
.
.
.

To bring this into context of `short_alloc`: 
Below we'll have `sizeof(alloc2) == sizeof(alloc8)`.

``` cpp
    {
        // TWO (2)
        short_alloc<uint8_t, 2*sizeof(uint8_t), alignof(uint64_t)>::arena_type ar2;
        short_alloc<uint8_t, 2*sizeof(uint8_t), alignof(uint64_t)> alloc2(ar2);

        // EIGHT (8)
        short_alloc<uint8_t, 8*sizeof(uint8_t), alignof(uint64_t)>::arena_type ar8;
        short_alloc<uint8_t, 8*sizeof(uint8_t), alignof(uint64_t)> alloc8(ar8);

        std::cout << "sizeof(alloc2) == " << sizeof(alloc2) << std::endl;
        std::cout << "sizeof(alloc8) == " << sizeof(alloc8) << std::endl;
        assert(sizeof(alloc2) == sizeof(alloc8)); // surprise: identical!
    }
```

The "culprit" is the compiler, and what it does with [this code in `short_alloc.h`](https://github.com/ajneu/short_alloc_experiments/blob/master/short_alloc.h#L32)

HowardHinnant commented:
I believe the rationale is so that you can have arrays of any type, and each element in the array will have the proper alignment.

ajneu commented:
Yes, but it still caught me off-guard recently, because if I have a struct with a single byte in it, then the compiler knows this. Why waste the ensuing 7 bytes? They could be used for multiple char variables. (At least, that's what I think)

Example:

``` cpp
struct AAA {
   alignas(8) char arr[1];
};

AAA myAAA;
char c2;   /* This could be placed at address &myAAA.arr[1], but it will never be so (right?)
              Instead it will will end up somewhere at &myAAA.arr[8]                 (right?)*/
char c3;
char c4;
char c5;
char c6;
char c7;
char c7;
```

But just realized something: are the addresses growing up or down??? (Might that is a reason for the behaviour?) If we're growing downwards, then maybe the rule is: get a new (completely unused) memory. Then it will be exactly the sizeof the alignment. Hmmm...

ajneu commented:
So maby the compiler will always waste it. And as a courtesy, it lets us know, by giving us that surprising `sizeof`. (??)

ajneu commented:
Returning to `short_alloc`...

> Currently I'm undecided if the best way to fix this bug is to allow the last allocation to be less than alignment (as you propose)

Well, I don't think we need to treat the last allocation specially.
Looking at the code ([link](https://github.com/ajneu/short_alloc_experiments/blob/master/short_alloc.h#L62)), we see that `ptr_` can indeed jump out beyond the array's end (`buf_ + N`), if the alignment is such that `n != aligned_n`. 

But even if `ptr_ > buf_ + N` who cares?
The code in the pull request will still work in all cases. (I think. I hope.)
But actually: (only) one little thing comes to mind: we might want to remove the function `used()` ([link](https://github.com/ajneu/short_alloc_experiments/blob/master/short_alloc.h#L45)), since that would really be confusing on the last allocation!

But somehow this brings us full circle, doesn't it? Because we can solve the problem of the `used()` function, by changing `arena` to use up a multiple of alignment. So that it wastes some bytes for the user! _(same as C++ compilers do, for overaligned arrays in structs)_
So maby that's the way to go?!!
Change [this line](https://github.com/ajneu/short_alloc_experiments/blob/master/short_alloc.h#L32) to:

``` cpp
    alignas(alignment) char buf_[(N+(alignment-1)) & ~(alignment-1)];
```

Ah...
but then we could also just `static_assert`, that `N` must be a multiple of alignment.
We force the user of `short_alloc`, to conform to:

``` cpp
    static_assert(N % alignment == 0, "N needs to be a multiple of alignment");
```

So I guess there might be a tradeoff:
- keep `used()` and `static_assert` in a manner that might force the user to "waste bytes" 
  or
- get rid of `used()` (here we don't waste bytes, but the user cannot call `used()`)
  or
- handle the last allocation specially (I actually never thought of this initially)

Thoughts?

HowardHinnant commented:
If we constrain `size` to be a multiple of `alignment`, then no space is ever wasted, and `ptr_` will always be within the bounds `[buf_, but_+N]`, and we don't have to worry about the last allocation being special.  Also, I'm not willing to give up this check:

```
assert(pointer_in_buffer(ptr_) && "short_alloc has outlived arena");
```

This life-time mistake between container and arena is too easy to make, especially with move semantics and swap.

ajneu commented:
Ah right, yes! You we're saying that all along.
I think it's good.

Thanks.

ajneu commented:
Took some time for me to understand your point properly! ;)
Sorry.

HowardHinnant commented:
You deserve the credit for bringing this issue to light.  Can you adjust your pull request with these two new static_asserts?  The one about the multiple, and the one that constrains the alignment to be less than or equal to max_align_t.

ajneu committed:
two static_asserts on alignment:
- size N needs to be a multiple of alignment Align
- operator new cannot handle alignments larger than alignof(std::max_align_t)

ajneu commented:
Ok, I've done I new commit. I hope it's correct.
(Note: I've include some other minor changes also, to allow it to compile without warnings when using flags `-Wall -Wextra -Werror`).

ajneu commented:
Oh hang on, it's missing some original code....

HowardHinnant commented:
No hurry.  I've got chicken wings I need to take care of. ;-)

ajneu commented:
Ah ok, diffing just between 3rd and 4th (lastest) commits had me confused. The accumulated diff under the "Files changes" tab looks correct, I believe
https://github.com/HowardHinnant/HowardHinnant.github.io/pull/4/files

Thanks.

HowardHinnant commented:
Looking good.  I appreciate the warning suppression and the wrapping of the static assert to reasonable line lengths.

On:
    `buf_ + N - ptr_ >= aligned_n`
vs:
   `ptr_ + aligned_n <= buf_ + N`

Algebraically these are the same statement.  And yours has better readability.  However we have a run-time invariant that `ptr_` is always in the range of `[buf_, buf_+N]`.  And a compile-time invariant that `buf_+N` never overflows.  And therefore `buf_+N-ptr_` is always in the range `[0, N]`.  But `aligned_n` might be just about anything.  Therefore I believe that yours is susceptible to overflow and mine is not.

Correctness trumps readability.

ajneu committed:
Prevent potential overflow (static_cast to suppress warnings)

ajneu commented:
Ah wow. Very instructive comment! Thanks.
Those corner cases are very interesting.

The compiler brings the following warning for
`buf_ + N - ptr_ >= aligned_n`
**->** 
`error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]`

See here: http://melpon.org/wandbox/permlink/iYUtP8EXEEVFAOoh

I suppose the **correct way** of handling it, while _eliminating the warning_ is:
`static_cast<decltype(aligned_n)>(buf_ + N - ptr_) >= aligned_n`
(where the left-hand-operand is cast to an unsigned: `std::size_t`)

The following would probably be **wrong**:
`buf_ + N - ptr_ >= static_cast<ptrdiff_t>(aligned_n)`
in that it would again just silence a warning, without solving the overflow problem.

I've made a 5th commit, handling it.

HowardHinnant commented:
Merged.
